File: stock-market-service/src/main/java/rs/edu/raf/strategy/StopLimitOrderStrategy.java
package rs.edu.raf.strategy;

import com.google.gson.Gson;
import org.springframework.amqp.rabbit.core.RabbitTemplate;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;
import rs.edu.raf.exceptions.StockDoesntExist;
import rs.edu.raf.model.*;
import rs.edu.raf.model.order.LimitOrder;
import rs.edu.raf.model.order.Order;
import rs.edu.raf.model.order.OrderAction;
import rs.edu.raf.model.order.StopLimitOrder;
import rs.edu.raf.repository.ListingOwnerRepository;
import rs.edu.raf.repository.OrderRepository;
import rs.edu.raf.repository.StockRepository;

@Component("stopLimitOrder")
public class StopLimitOrderStrategy extends AbstractOrderProcessStrategy {
    private StockRepository stockRepository;
    private OrderRepository orderRepository;

    public StopLimitOrderStrategy(@Value("${update.balance.exchange}") String UPDATE_BALANCE_EXCHANGE,
                                      @Value("${update.balance.routing.key}") String UPDATE_BALANCE_ROUTING_KEY, OrderRepository orderRepository, ListingOwnerRepository listingOwnerRepository, RabbitTemplate rabbitTemplate, Gson gson, StockRepository stockRepository) {
        super(UPDATE_BALANCE_EXCHANGE, UPDATE_BALANCE_ROUTING_KEY, orderRepository, listingOwnerRepository, rabbitTemplate, gson);
        this.stockRepository = stockRepository;
        this.orderRepository = orderRepository;
    }

    @Override
    public void processOrder(Order order) {
        if(!isOrderExecutable(order)) {
            return;
        }

        StopLimitOrder stopLimitOrder = (StopLimitOrder) order;
        orderRepository.save(createLimitOrder(stopLimitOrder));
        orderRepository.delete(stopLimitOrder);
    }

    @Override
    public boolean isOrderExecutable(Order order) {
        String ticker = order.getTicker();

        Stock stock = stockRepository.findById(ticker).orElseThrow(()->new StockDoesntExist(ticker));

        StopLimitOrder stopLimitOrder = (StopLimitOrder) order;

        return (stock.getLow() < stopLimitOrder.getStop() && stopLimitOrder.getOrderAction() == OrderAction.SELL) ||
                (stock.getHigh() > stopLimitOrder.getStop() && stopLimitOrder.getOrderAction() == OrderAction.BUY);
    }

    private LimitOrder createLimitOrder(StopLimitOrder stopLimitOrder) {
        LimitOrder limitOrder = new LimitOrder();

        limitOrder.setLimitAmount(stopLimitOrder.getLimitAmount());
        limitOrder.setOrderAction(stopLimitOrder.getOrderAction());
        limitOrder.setQuantity(stopLimitOrder.getQuantity());
        limitOrder.setTicker(stopLimitOrder.getTicker());
        limitOrder.setOrderCreator(stopLimitOrder.getOrderCreator());

        return limitOrder;
    }
}

File: stock-market-service-kotlin/src/main/kotlin/org/example/stockmarketservicekotlin/strategy/StopLimitOrderStrategy.kt
package org.example.stockmarketservicekotlin.strategy

import com.google.gson.Gson
import org.example.stockmarketservicekotlin.exceptions.StockDoesntExistException
import org.example.stockmarketservicekotlin.model.Stock
import org.example.stockmarketservicekotlin.model.order.LimitOrder
import org.example.stockmarketservicekotlin.model.order.Order
import org.example.stockmarketservicekotlin.model.order.OrderAction
import org.example.stockmarketservicekotlin.model.order.StopLimitOrder
import org.example.stockmarketservicekotlin.repository.ListingOwnerRepository
import org.example.stockmarketservicekotlin.repository.OrderRepository
import org.example.stockmarketservicekotlin.repository.StockRepository
import org.springframework.amqp.rabbit.core.RabbitTemplate
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component

@Component("stopLimitOrder")
class StopLimitOrderStrategy(
    @Value("\${update.balance.exchange}") UPDATE_BALANCE_EXCHANGE: String,
    @Value("\${update.balance.routing.key}") UPDATE_BALANCE_ROUTING_KEY: String,
    rabbitTemplate: RabbitTemplate,
    gson: Gson,
    orderRepository: OrderRepository,
    private val stockRepository: StockRepository,
    listingOwnerRepository: ListingOwnerRepository,
) : AbstractOrderProcessStrategy(
    UPDATE_BALANCE_EXCHANGE, UPDATE_BALANCE_ROUTING_KEY, orderRepository, listingOwnerRepository, rabbitTemplate, gson
) {
    override fun processOrder(order: Order) {
        if (!isOrderExecutable(order)) {
            return
        }

        val stopLimitOrder: StopLimitOrder = order as StopLimitOrder
        orderRepository.save(createLimitOrder(stopLimitOrder))
        orderRepository.delete(stopLimitOrder)
    }

    public override fun isOrderExecutable(order: Order): Boolean {
        val ticker: String = order.ticker

        val stock: Stock = stockRepository.findById(ticker).orElseThrow { StockDoesntExistException(ticker) }

        val stopLimitOrder: StopLimitOrder = order as StopLimitOrder

        return (stock.low < stopLimitOrder.stop && stopLimitOrder.orderAction == OrderAction.SELL) ||
                (stock.high > stopLimitOrder.stop && stopLimitOrder.orderAction == OrderAction.BUY)
    }

    private fun createLimitOrder(stopLimitOrder: StopLimitOrder): LimitOrder {
        return LimitOrder(
            null,
            stopLimitOrder.orderAction,
            stopLimitOrder.ticker,
            stopLimitOrder.quantity,
            stopLimitOrder.orderCreator,
            stopLimitOrder.limitAmount
        )
    }
}